### DISCUSSION TITLE
Can I get DataContext or DataConnection from IQueryable？

### DISCUSSION THREAD

**Discussion by darkflame0** [linq2db/linq2db#3628] [Q&A] [answered]
I want to  create an ExtensionMethod using recursion cte on IQueryable,but I cant call GetCte without DataContext.

otherwise,Can I LoadWith a tree with cte?

**Comment by MaceWindu** [comment]
You can try this to get `IDataContext`, which usually could be casted to `DataContext` or `DataConnection` (if `IDataContext` is not enough)
```cs
((dynamic)queryable.Provider).DataContext
```

As `IQueryProvider` implementation is not public class, you need to use dynamic or reflection (https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/Linq/ExpressionQuery.cs#L27)

**Comment by sdanyliv** [comment, marked as answer]
You can use `LinqToDB.Linq.Internals.GetDataContext(query)`
